editor/zls/src/document_symbols.rs

```diff
@@ -0,0 +1,150 @@
+        ResponseError::new(ErrorCode::REQUEST_FAILED, format!("parsing error: {err}"))
+    })?;
+
+    let symbols = ast
+        .inner
+        .declarations
+        .iter()
+        .flat_map(|decl| declaration_to_symbol(decl, document))
+        .collect();
+    Ok(Some(DocumentSymbolResponse::Nested(symbols)))
+}
+
+fn declaration_to_symbol(
+    decl: &Sp<DeclSymbol<Declaration>>, document: &FullTextDocument,
+) -> Vec<DocumentSymbol> {
+    match &decl.inner.inner {
+        Declaration::Module(module_def) => {
+            let children = module_def
+                .declarations
+                .iter()
+                .flat_map(|decl| declaration_to_symbol(decl, document))
+                .collect();
+
+            match &module_def.name {
+                Some(name_ref) => {
+                    vec![name_to_symbol(
+                        name_ref,
+                        decl.span(),
+                        document,
+                        SymbolKind::MODULE,
+                        children,
+                    )]
+                }
+                None => children,
+            }
+        }
+        Declaration::UseDef(_) => vec![],
+        Declaration::Data(data_def) => {
+            let children =
+                data_def.ctors.iter().map(|ctor| ctorv_to_symbol(&ctor.ctorv, document)).collect();
+
+            vec![name_to_symbol(&data_def.name, decl.span(), document, SymbolKind::ENUM, children)]
+        }
+        Declaration::Codata(codata_def) => {
+            let children = codata_def
+                .dtors
+                .iter()
+                .map(|dtor| dtorv_to_symbol(&dtor.dtorv, document))
+                .collect();
+
+            vec![name_to_symbol(
+                &codata_def.name,
+                decl.span(),
+                document,
+                SymbolKind::CLASS,
+                children,
+            )]
+        }
+        Declaration::Alias(alias_def) => {
+            // TODO: lookup symbol to display correct kind
+            vec![name_to_symbol(&alias_def.name, decl.span(), document, SymbolKind::ENUM, vec![])]
+        }
+        Declaration::Define(def) => {
+            // TODO: handle local definitions
+            let children = vec![];
+
+            let kind = match def.0.params.len() {
+                0 => SymbolKind::CONSTANT,
+                _ => SymbolKind::FUNCTION,
+            };
+
+            vec![name_to_symbol(&def.0.name.0, decl.span(), document, kind, children)]
+        }
+        // TODO: handle main, local definitions
+        Declaration::Main(_) => vec![],
+    }
+}
+
+#[allow(deprecated)]
+fn name_to_symbol(
+    name: &impl NameT, item_span: &Span, document: &FullTextDocument, kind: SymbolKind,
+    children: Vec<DocumentSymbol>,
+) -> DocumentSymbol {
+    let name_view = name.name();
+    let name_str = name_view.ident.inner.clone();
+    let name_range = span_to_range(name_view.ident.span(), document);
```

Comment:
Would you suggest that we change `Span` to `Range` in the first place?

Reply:
I'm not exactly sure what you're thinking of here, but I wouldn't recommend changing `Span`. (I'm also looking into a way to abstract away this explicit conversion.)

Reply:
In case it's what you're thinking, `Range` here isn't the standard library `Range`; this is a type [specified in the LSP](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range), and provided in Rust by `lsp_types`. It's an awkward type, so I'd recommend avoiding it to the extent possible, though it must be used to communicate with the LSP client.

For some extra context, a theme of LSP is unfortunate design choices as a result of VS Code (an Electron app written in JS), backwards compatibility, and assumptions only valid for "popular", usually imperative, languages (e.g. JS/Python/Java/...). In this case, it's mostly the first one:
- LSP's `Range` is `{ start: Position, end: Position }`
- LSP's [`Position`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position) is `{ line: uint, character: uint }`, where `line` is a (0-indexed) line number, and `character` is a (0-indexed) offset into that line
- The units of `character` are UTF-16 surrogate pairs (because JS). Rust encodes strings as UTF-8, so UTF-8 byte offsets, or even Unicode codepoint offsets, would be much more natural; UTF-16 is very inconvenient.
- For e.g. ASCII, these all agree; JS `"a".length` and Rust `"a".len()` are both `1`. However, this does not hold for all characters; JS `"🤗".length` is `2`, while Rust `"🤗".len()` is `4`.
- (Recent versions of LSP allow the client and server to agree on another encoding, like UTF-8. However, not all clients support UTF-8 offsets, and the unique encoding required to be supported is UTF-16. For simplicity, I didn't implement this here.)
- [Bad things happen when you don't handle this properly.](https://fasterthanli.me/articles/the-bottom-emoji-breaks-rust-analyzer)

This means that `Range` is based on UTF-16 offsets, which is not usually useful, and would be a poor representation for the compiler. Since Rust operates over UTF-8, we are effectively required to perform a conversion somewhere, so there won't really be a way around it.

Reply:
I never imagined `Range` to be such a terrible encoding 😱 Good thing you've figured it out. Thank you for such a detailed explanation!